packages/storage/exp/index.ts

```diff
@@ -53,7 +53,36 @@ export { ListOptions, ListResult } from '../src/list';
 export { UploadTask } from '../src/task';
 export { UploadResult, UploadTaskSnapshot } from '../src/tasksnapshot';
 export { StringFormat } from '../src/implementation/string';
-
+export {
+  RequestInfo as _RequestInfo,
+  UrlParams as _UrlParams
+} from '../src/implementation/requestinfo';
+export { Request as _Request } from '../src/implementation/request';
+export { Location as _Location } from '../src/implementation/location';
+export { XhrIoPool as _XhrIoPool } from '../src/implementation/xhriopool';
+export {
+  TaskEvent,
+  TaskState,
+  InternalTaskState as _InternalTaskState
```

Comment:
We should not export internal things. I commented some of them out and api extractor continues to work without giving any warnings. 

If I remember correctly, api extractor only asks you to export an symbol if it is used in a exported function/class/etc., but itself not being exported. It means we are referencing internal symbols in the public interface, which we should fix.
What is the specific warning that asks you to export internal symbols?

Reply:
I removed all the internal exports and ran api-report again, you can view the md file now to see the warning comments, an example is:

```
// Warning: (ae-forgotten-export) The symbol "Request" needs to be exported by the entry point index.d.ts
```

Reply:
Right. These symbols are referenced in the public classes/interfaces/functions, so they need to be exported even if the method that references them are annotated as `internal`. Here, `internal` has a slightly different meaning than just "hidden from public". It is designed to be used in a project with many packages that depend on each other. When a symbol is annotated as `internal` in a package, it means the symbol will be used in other packages in the same project, but not for public use. 

It means that api-extractor will generate 2 different d.ts files:
- a dev version of the d.ts for local development, which has all public APIs +  internal APIs/types (annotated as `internal`), which other internal packages can see.
- a public version of the d.ts, where all `internal` typings are stripped
`@firebase/app-exp` uses this feature to make, for example, `_registerComponent()` available for other firebase packages, but hide it from developers.

Since both d.ts are generated from the same source file, you need to export everything including the internal ones, so the warning makes sense.
In your case, you don't really need the `internal` feature because no other internal packages need to access any non-public APIs from storage.

What we should do is create a public interface for every implementation classes (e.g. StorageService, Reference, etc) in the storage-types package, and only reference these public interfaces in the public APIs. This way, you don't have to annotate things with `internal` in the implementation classes and there should be no warning to ask you to export internal things.

Reply:
> What we should do is create a public interface for every implementation classes (e.g. StorageService, Reference, etc) in the storage-types package, and only reference these public interfaces in the public APIs. This way, you don't have to annotate things with `internal` in the implementation classes and there should be no warning to ask you to export internal things.

Does that not contradict our plan to move away from manual types? I thought of one the goals of API Extractor was to not have to do this.

Reply:
Not really. We want to have consistent typings by getting rid of the manually curated `firebase/index.d.ts`, so there is a single source of truth for the public APIs for every SDK, which is `@firebase/foo` + `@firebase/foo-types`. `@firebase/foo` defines what gets exported and `@firebase/foo-types` defines the public types, e.g. parameter types and return types.

API extractor can help us get rid of `@firebase/foo-types` as well as manually defined public types via its release tag feature, but I think we want to keep `@firebase/foo-types` to allow people to depend on our public types without depending on the actual implementation, which is, IIUC, why we structured our packages this way back in the day.

Reply:
Just to be clear before I go through all the methods, I should be doing something like this for every public method? (doing it for `uploadBytes()` in `packages/storage/exp/index.ts`).

```
import * as types from '@firebase/storage-types/exp';
import { Metadata } from '../src/metadata';
import { UploadResult } from '../src/tasksnapshot';
import {
  uploadBytes as uploadBytesInternal,
  Reference
} from '../src/reference';

export function uploadBytes(
  ref: types.StorageReference,
  data: Blob | Uint8Array | ArrayBuffer,
  metadata: types.Metadata
): Promise<UploadResult> {
  return uploadBytesInternal(ref as Reference, data, metadata as Metadata);
}
```

Seems like a lot of boilerplate to repeat, but if it's the right way to do things, I'm sure it needs to be done.

Also is packages/storage-types/exp/ folder the best place to put it? Or should I make a packages-exp/storage-types-exp? In the first case, will I have to do something special to make sure exp users get the package since @firebase/storage-types doesn't normally get published as part of exp release?

Reply:
Exactly. Please make sure the implementations actually implement these interfaces to keep them in sync. Storage-types/exp sounds good to me, just make sure the new file will be included when publishing.

Reply:
I've done this and everything seems fine on the code side, the api-extractor md file looks clean, but the problem is that when I generate the reference docs, all the types from index.d.ts (like types.Metadata) are dead-ends and do not have a link with details. For example, the reference doc page for updateMetadata(), a function which has types.Metadata as a param and return type, should be linking to a separate page documenting the Metadata interface, I think. But if the type was imported from @firebase/storage-types/exp it's just a text link, the user cannot find any info on Metadata's param. Same with UploadTask which is even more important I think.

Reply:
These types should link to @firebase/storage-types/exp correctly when we generate reference docs for all packages together. See the return type of `getApp()` which links to a page in `@firebase/app-types`.

You need to create a api-extractor script in `@firebase/storage-types`, then run `yarn docgen:exp` at repo root. `docgen:exp` will call this script in all projects and copy their output to a folder and generate reference docs for all of them with linking across packages. You also need to make sure the package names match in the api-extractor output, so linking works. There might be some special handling you need to do to get matching package names, since you are rewriting it in place.